ampd/src/queue/queued_broadcaster.rs

```diff
@@ -78,113 +79,99 @@ where
                 },
+                _ = interval.tick() => self.broadcast_all().await?.then(|_| {interval.reset()}),
             }
         }
 
-        broadcast_all(&mut queue, &mut broadcaster).await?;
+        self.broadcast_all().await?;
 
         Ok(())
     }
 
     pub fn client(&self) -> QueuedBroadcasterClient {
         QueuedBroadcasterClient {
-            sender: self.channel.0.clone(),
+            sender: self
+                .channel
+                .as_ref()
+                .expect("channel must be set before running")
+                .0
+                .clone(),
         }
     }
-}
 
-async fn handle_msg<T>(
-    batch_gas_limit: u64,
-    queue: &mut MsgQueue,
-    broadcaster: &mut T,
-    interval: &mut time::Interval,
-    msg_and_res_chan: MsgAndResChan,
-) -> Result
-where
-    T: Broadcaster,
-{
-    let (msg, tx) = msg_and_res_chan;
+    async fn broadcast_all(&mut self) -> Result {
+        let msgs = self.queue.pop_all();
 
-    match broadcaster.estimate_fee(vec![msg.clone()]).await {
-        Ok(fee) => {
-            tx.send(Ok(())).map_err(|_| Report::new(Error::Client))?;
+        match msgs.len() {
+            0 => Ok(()),
+            n => {
+                info!(message_count = n, "ready to broadcast messages");
 
-            if fee.gas_limit.saturating_add(queue.gas_cost()) >= batch_gas_limit {
-                warn!(
-                    queue_size = queue.len(),
-                    queue_gas_cost = queue.gas_cost(),
-                    "exceeded batch gas limit. gas limit can be adjusted in ampd config"
-                );
-                broadcast_all(queue, broadcaster).await?;
-                interval.reset();
+                self.broadcaster
+                    .broadcast(msgs)
+                    .await
+                    .map(|_| ())
+                    .change_context(Error::Broadcast)
             }
-
-            let message_type = msg.type_url.clone();
-            queue
-                .push(msg, fee.gas_limit)
-                .change_context(Error::Queue)?;
-            info!(
-                message_type,
-                queue_size = queue.len(),
-                queue_gas_cost = queue.gas_cost(),
-                "pushed a new message into the queue"
-            );
-        }
-        Err(err) => {
-            tx.send(Err(err).change_context(Error::EstimateFee))
-                .map_err(|_| Report::new(Error::Client))?;
         }
     }
 
-    Ok(())
-}
-
-async fn broadcast_all<T>(queue: &mut MsgQueue, broadcaster: &mut T) -> Result
-where
-    T: Broadcaster,
-{
-    let msgs = queue.pop_all();
-
-    match msgs.len() {
-        0 => Ok(()),
-        n => {
-            info!(message_count = n, "ready to broadcast messages");
-
-            broadcaster
-                .broadcast(msgs)
-                .await
-                .map(|_| ())
-                .change_context(Error::Broadcast)
+    async fn handle_msg(
```

Comment:
looks much better now. Could we return `Result<Interval,_>` instead? I'm not a fan of silently modifying the interval as a side effect